File: Piece of cake/222. Count Complete Tree Nodes/main.cpp
#include <bits/stdc++.h>
using namespace std;
// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
   public:
    int countNodes(TreeNode* root) {
        int dep = 0;
        TreeNode* cur = root;
        while (cur) {
            cur = cur->left;
            ++dep;
        }
        auto chk = [&](int n) {
            TreeNode* cur = root;
            for (int mask = 1 << (dep - 2); cur != nullptr && mask > 0; mask >>= 1)
                cur = (mask & n) ? cur->right : cur->left;
            return cur != nullptr;
        };
        int hi = 1 << dep, lo = hi >> 1;
        while (hi - lo > 1) {
            int mid = lo + (hi - lo) / 2;
            if (chk(mid)) lo = mid;
            else hi = mid;
        }
        return lo;
    }
};

int main() { return 0; }

File: Piece of cake/222. Count Complete Tree Nodes/main.rs
#![allow(dead_code)]
struct Solution;
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None,
    }
  }
}
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
  pub fn count_nodes(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
    let dep = {
      let mut dep = 0;
      let mut cur = root.clone();
      while let Some(nxt) = cur.take() {
        cur = nxt.borrow().left.clone();
        dep += 1;
      }
      dep
    };
    let chk = |n| {
      let mut cur = root.clone();
      let mut mask = 1 << (dep - 2);
      while cur.is_some() && mask > 0 {
        if mask & n == 0 {
          cur = cur.take().borrow().left.clone();
        } else {
          cur = cur.take().borrow().right.clone();
        }
        mask >>= 1;
      }
      cur.is_some()
    };
    let mut hi = 1 << dep;
    let mut lo = hi >> 1;
    while hi - lo > 1 {
      let mid = lo + (hi - lo) / 2;
      if chk(mid) {
        lo = mid;
      } else {
        hi = mid;
      }
    }
    lo
  }
}

fn main() {}
